tensorflow-framework/src/main/java/org/tensorflow/framework/data/impl/MapOptional.java

```diff
@@ -0,0 +1,42 @@
+/*
+ * Copyright 2020 The TensorFlow Authors. All rights reserved.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+package org.tensorflow.framework.data.impl;
+
+import org.tensorflow.DataType;
+import org.tensorflow.Operand;
+import org.tensorflow.framework.data.DatasetOptional;
+import org.tensorflow.tools.Shape;
+
+import java.util.List;
+import java.util.function.Function;
+
+public class MapOptional extends DatasetOptional {
+  private final Function<List<Operand<?>>, List<Operand<?>>> mapper;
+
+  public MapOptional(
+      DatasetOptional source,
+      Function<List<Operand<?>>, List<Operand<?>>> mapper,
+      List<DataType<?>> outputTypes,
+      List<Shape> outputShapes) {
+    super(source.getOpsInstance(), source.getOptionalVariant(), outputTypes, outputShapes);
```

Comment:
I refactored this constructor to take a `DatasetOptional`.

Since these constructors are in the `org.tensorflow.data.impl` package, but are constructed in `org.tensorflow.data.Dataset`, I don't think the constructors can trivially be made package-private...